[Sol->Yul] Implementing bytes concat (#11011)
ethereum/solidity: docs/types/reference-types.rst

    that you are accessing the low-level bytes of the UTF-8 representation,
    and not the individual characters.

> You can concatenate variable number of ``bytes`` or ``bytes1 ... bytes32`` using ``bytes.concat``.

Reviewer: Can you link this from the "Fixed Bytes" section?

> It returns single ``bytes memory`` array.

Reviewer: ```suggestion
The function returns single ``bytes memory`` array that contains the contents of the arguments without padding.
```

> If you want to use string parameter, you need to explicitly convert it to ``bytes``.

Reviewer: ```suggestion
If you want to use string parameters or other types, you need to convert them to ``bytes`` or ``bytes1``/.../``bytes32`` first.
```

::

    // SPDX-License-Identifier: GPL-3.0

>     pragma solidity >=0.8.3 < 0.9.0;

Reviewer: ```suggestion
    pragma solidity ^0.8.3;
```

Author: I needed to use `pragma solidity > 0.8.2;` in order for `cmdLineTests` to pass. I think that is fine

    contract C {
        bytes s = "Storage";
        function f(bytes calldata c, string memory m, bytes16 b) public pure {

>             bytes memory a = bytes.concat(s, c, "Literal", bytes(m), b);

Reviewer: Do calldata slices work?

Author: Should work, I'll add that into example, and will also add semantic tests for `calldata slices`

Author: Added slice in docs example, as well as in semantic tests now

            assert((s.length + c.length + 7 + bytes(m).length + 16) == a.length);
        }
    }

> If you call ``bytes.concat`` without parameters it will return zero sized ``bytes``.

Reviewer: ```suggestion
If you call ``bytes.concat`` without arguments it will return an empty ``bytes`` array.
```

    that you are accessing the low-level bytes of the UTF-8 representation,
    and not the individual characters.

> .. index:: ! bytes-concat

Reviewer: The first paragraph of this section still sits `abi.encodePacked` for concatenation.

Author: Changed to `bytes.concat(bytes(s1), bytes(s2))`. Also added a test for this particular case as well. We didn't have the one with strings converted to bytes.

> .. _bytes-concat:

Reviewer: Can you add a section heading?

Author: Added

    that you are accessing the low-level bytes of the UTF-8 representation,
    and not the individual characters.

.. index:: ! bytes-concat

.. _bytes-concat:

``bytes.concat`` function
^^^^^^^^^^^^^^^^^^^^^^^^^

> You can concatenate variable number of ``bytes`` or ``bytes1 ... bytes32`` using ``bytes.concat``.
> The function returns single ``bytes memory`` array that contains the contents of the arguments without padding.
> If you want to use string parameters or other types, you need to convert them to ``bytes`` or ``bytes1``/.../``bytes32`` first.

Reviewer: ```suggestion
You can concatenate a variable number of ``bytes`` or ``bytes1 ... bytes32`` using ``bytes.concat``.
The function returns a single ``bytes memory`` array that contains the contents of the arguments without padding.
If you want to use string parameters or other types, you need to convert them to ``bytes`` or ``bytes1``/.../``bytes32`` first.
```
